packages/core/src/node/core-process-manager.ts

```diff
@@ -0,0 +1,98 @@
+/********************************************************************************
+ * Copyright (C) 2021 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import * as cp from 'child_process';
+import { injectable } from 'inversify';
+
+@injectable()
+export class CoreProcessManager {
+
+    /**
+     * Sends SIGTERM to a process and its children recursively.
+     *
+     * @param ppid The process to kill along with its children (recursive).
+     */
+    terminateProcessTree(ppid: number): void {
+        if (process.platform === 'win32') {
+            this.win_taskkillTree(ppid);
+        } else {
+            if (ppid === this.unix_getpgid(ppid)) {
+                // When a process pgid === pid this means the the process is a group leader.
+                // We can then kill every process part of its group by doing `kill(-pgid)`.
+                process.kill(-ppid);
+            } else {
+                // If the process is not a group leader, we will resort to listing the process hierarchy manually:
+                for (const pid of this.unix_getChildrenRecursive(ppid)) {
+                    // Prevent killing the current process:
+                    if (pid !== process.pid) {
+                        // Don't stop if a process fails to be killed (keep on killing the others):
+                        try {
+                            process.kill(pid);
+                        } catch (error) {
+                            console.error(error);
+                        }
+                    }
+                }
+                process.kill(ppid);
+            }
+        }
+    }
+
+    protected unix_getpgid(pid: number): number {
+        const { stdout } = this.spawnSync('ps', ['-p', pid.toString(10), '-o', 'pgid=']);
+        return Number.parseInt(stdout, 10);
+    }
+
+    protected unix_getChildrenRecursive(ppid: number): Set<number> {
```

Comment:
I'm still not sure about this implementation and the contingency in which it would kick in. It appears that PPID is more volatile than GID: a process whose parent is killed changes its PPID, but it appears that the GID remains the same. That means that a system that switches to PPID when it detects a disruption in GID may be switching to an even less reliable option.

What are the scenarios in which you think that this system will kick in, and what results do you expect?

Reply:
This is only used when the backend could not be started with its own process group, or isn't the group leader. About following PPIDs to find children, this is what ps-tree does, roughly. The PPID only ever changes from its original one to 1 (`init`) AFAIK.

But I now think that we should always parse the hierarchical process tree in case some child create their own group (and then keep running once we kill the backend process group). I will make the update.

Reply:
Hmm... So there are two times when this code would be useful:
- If somehow the GID of the backend main process is not the PID of the backend main process. As you say, that may be because we couldn't start the backend main process as the group leader for some reason. Then it's true that finding the children may be the best option, even if it may be fragile because children who have been adopted by `init` will escape.
- If some backend process creates its own group, and so escapes the GID-based approach. In that case, tracking children manually (assuming the tree is unbroken) is more reliable than the GID.

There are tradeoffs: assuming GID=PPID, then GID is going to be more reliable, because even processes adopted by `init` will be caught, unless some child made deliberate moves to escape the group.

Are we interested in covering cases in which some child process may not have managed its own children properly, leaving them to be adopted by `init`? Switching to a pure PPID-based approach will always miss those.

Reply:
I agree with both points, and I updated the current changes to kill based on PGID and following parent/child relationships.